config/credhub_secret_test.go

```diff
@@ -0,0 +1,147 @@
+package config_test
+
+import (
+	"code.cloudfoundry.org/cli/plugin/pluginfakes"
+	"errors"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/config"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/httpclient/httpclientfakes"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/serviceutil/serviceutilfakes"
+)
+
+var _ = Describe("CredhubSecret", func() {
+
+	const (
+		accessToken       = "fake-access-token"
+		bearerAccessToken = "bearer fake-access-token"
+		serviceURI        = "service-uri/"
+		secretsURI        = "service-uri/secrets"
+		configServerName  = "fake-config-server-name"
+	)
+
+	var (
+		fakeCliConnection *pluginfakes.FakeCliConnection
+		credhubSecret     config.CredHubSecret
+		fakeAuthClient    *httpclientfakes.FakeAuthenticatedClient
+		fakeResolver      *serviceutilfakes.FakeServiceInstanceResolver
+		secretsError      error
+	)
+
+	BeforeEach(func() {
+		fakeCliConnection = &pluginfakes.FakeCliConnection{}
+		fakeAuthClient = &httpclientfakes.FakeAuthenticatedClient{}
+		fakeResolver = &serviceutilfakes.FakeServiceInstanceResolver{}
+
+		fakeCliConnection.AccessTokenReturns(bearerAccessToken, nil)
+		fakeResolver.GetServiceInstanceUrlReturns(serviceURI, nil)
+	})
+
+	JustBeforeEach(func() {
+		credhubSecret = config.NewCredHubSecret(fakeCliConnection, fakeAuthClient, fakeResolver)
+	})
+
+	Describe("CredHub Add Secret", func() {
+
+		JustBeforeEach(func() {
+			secretsError = credhubSecret.Add(configServerName, "application/cloud/one", "{\"key\":\"secret\"}")
+		})
+
+		Context("when add returns success", func() {
+
+			BeforeEach(func() {
+				fakeAuthClient.DoAuthenticatedPutReturns(200, nil)
+			})
+
+			It("should call secrets endpoint", func() {
+				Expect(secretsError).To(BeNil())
+				Expect(fakeAuthClient.DoAuthenticatedPutCallCount()).Should(Equal(1))
+
+				url, bodyType, body, token := fakeAuthClient.DoAuthenticatedPutArgsForCall(0)
+				Expect(url).To(Equal(secretsURI + "/application/cloud/one"))
+				Expect(token).To(Equal(accessToken))
+				Expect(bodyType).To(Equal("application/json"))
+				Expect(body).To(Equal("{\"key\":\"secret\"}"))
+			})
+		})
+
+		Context("when add fails", func() {
+
+			var e = errors.New("failed to add secret or is not supported by this config server")
+
+			BeforeEach(func() {
+				fakeAuthClient.DoAuthenticatedPutReturns(500, e)
+				secretsError = credhubSecret.Add("fake-service-name", "application/cloud/one", "{\"key\":\"secret\"}")
+			})
+
+			It("should return error message", func() {
+				Expect(secretsError).To(Equal(e))
+			})
+		})
+
+		Context("when add calls an old version of SCS", func() {
+
+			var e = errors.New("failed to add secret or is not supported by this config server")
```

Comment:
Can we tell when we're talking to an old config server instance? It would be nice if we could tell the 'not supported' case from the 'failed to add' case. No worries if not though.

Reply:
Sadly, we get a 403 when calling the `/secrets/` endpoint in an old config server instance.
The only solution to give a better name is to query the broker version somehow, but I thought that was a bit of an overkill and we are following this approach in other areas of the code. What do you think?

Reply:
Yep agreed, I don't think it's worth querying the broker - was just wondering if this was possible with the info we have. Do we also get a 403 if we call the endpoint in a new config server in some cases? Maybe if we call an instance that the user is not a space developer for?

There's actually a test for v2 vs v3 going on in code you're already calling in `service_instance_resolver`... Up to you if you wanted to refactor to make use of that, but in any case I don't think this is anything that should prevent this from being merged - more just curious :)

Reply:
It is a valid point, asked @csterwa explicitly about that. We could enhance afterwards if we feel like it is important.